### PR TITLE
[#159] [Generate project with Mason] Generate the new project & keep the generated project in the repository root folder for reference

### PR DESCRIPTION
**Pull request by luongvo**
- Solves #159

## What happened 👀

Add a new workflow on the `develop` branch at the root template project to 
  - automatically generate the new project.
  - commit & push the generated project to the repository root folder under `/sample`.

## Insight 📝

- To ensure the new `sample` project content should be updated accordingly, remove the `sample` project before generating its new content.

## Proof Of Work 📹

- The new `sample` project content is updated accordingly https://github.com/nimblehq/flutter-templates/actions/runs/4970597107/jobs/8894565147
  - [1] non-existing files in the new revision are deleted ✅ 
  - [2] out-of-date content files are overwritten ✅ 
  - [3] the same file content with a new name should be renamed/moved ✅ 
  - the rest are added ✅ 

  - Test commit with old changes https://github.com/nimblehq/flutter-templates/pull/196/commits/7d1c5f9136d88e98e806faadb1e145905e5e7d35

    image

  - Update commit: https://github.com/nimblehq/flutter-templates/pull/196/commits/f49111d3244a552cb4e575e9c337d15366f6cfab

    image

    image

- A `sample` project folder is added
  
  image

  image

### CODE REVIEW

File: .github/workflows/generate_sample.yml [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    branches:
+      - develop
+
+jobs:
+  test:
+    name: Generate sample project
+    runs-on: ubuntu-latest
+    timeout-minutes: 30
+    steps:
+      - name: Checkout
```
Comment (chornerman) [inline .github/workflows/generate_sample.yml:14]: (Optional)
```suggestion
      - name: Check out
```

**Review by chornerman — APPROVED** [review]
rest LGTM 🚀

File: .github/workflows/generate_sample.yml [hunk 2]
```diff
@@ -0,0 +1,42 @@
+name: Generate sample project
```
Comment (manh-t) [inline .github/workflows/generate_sample.yml:1]:
```suggestion
name: Generate a sample project
```

File: .github/workflows/generate_sample.yml [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+      - name: Checkout
+        uses: actions/checkout@v3
+
+      - name: Set up Flutter environment
+        uses: subosito/flutter-action@v2
+        with:
+          channel: 'stable'
+          flutter-version: '3.3.10'
+
+      - name: Remove sample project
```
Comment (manh-t) [inline .github/workflows/generate_sample.yml:23]:
```suggestion
      - name: Remove the sample project
```

File: .github/workflows/generate_sample.yml [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        run: |
+          rm -rf sample
+
+      - name: Generate new sample project
```
Comment (manh-t) [inline .github/workflows/generate_sample.yml:27]:
```suggestion
      - name: Generate a new sample project
```

File: .github/workflows/generate_sample.yml [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+        run: |
+          dart pub global activate mason_cli
+          mason get
+          mason make template -c mason-config.json
+          # Move the generated project to the root directory for next steps & cleanup to not affect static code analysis
+          rsync -av --remove-source-files flutter_templates/ ./sample
+          rm -rf bricks/template
```
Comment (manh-t) [inline .github/workflows/generate_sample.yml:34]: @luongvo why don't we update the project name in `mason-config.json` to `sample`? Then we don't need the cleaning commands.
```suggestion
          mason make template -c mason-config.json
```

Comment (manh-t) [reply .github/workflows/generate_sample.yml:34]: Mason generates a new project called `sample` in the root directory 😄 
image

File: .github/workflows/generate_sample.yml [hunk 6]
```diff
@@ -0,0 +1,42 @@
+name: Generate sample project
+
+on:
+  push:
+    branches:
+      - develop
+
+jobs:
+  test:
+    name: Generate sample project
```
Comment (markgravity) [inline .github/workflows/generate_sample.yml:10]:
```suggestion
    name: Generate a sample project
```

File: .github/workflows/generate_sample.yml (hunk 1 above)
Comment (luongvo) [reply .github/workflows/generate_sample.yml]: Fixed in https://github.com/nimblehq/flutter-templates/pull/196/commits/be22efc9cd04ee3e2bf3bb8d95bc1e13a148e5ef

File: .github/workflows/generate_sample.yml (hunk 2 above)
Comment (luongvo) [reply .github/workflows/generate_sample.yml]: Fixed in https://github.com/nimblehq/flutter-templates/pull/196/commits/be22efc9cd04ee3e2bf3bb8d95bc1e13a148e5ef

File: .github/workflows/generate_sample.yml (hunk 6 above)
Comment (luongvo) [reply .github/workflows/generate_sample.yml]: Fixed in https://github.com/nimblehq/flutter-templates/pull/196/commits/be22efc9cd04ee3e2bf3bb8d95bc1e13a148e5ef

File: .github/workflows/generate_sample.yml (hunk 5 above)
Comment (luongvo) [reply .github/workflows/generate_sample.yml]: @manh-t I don't want to change the `project_name` in the config publicly as I want to keep it as default options. Hence, I added a step to change the field in CD workflow instead https://github.com/nimblehq/flutter-templates/pull/196/commits/cc50c5b5993e8f9c560cd077b0f3f120a555fd09

File: .github/workflows/generate_sample.yml [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    timeout-minutes: 30
+    steps:
+      - name: Check out
+        uses: actions/checkout@v3
+
+      - name: Set up Flutter environment
+        uses: subosito/flutter-action@v2
+        with:
+          channel: 'stable'
+          flutter-version: '3.3.10'
```
Comment (nmint8m) [inline .github/workflows/generate_sample.yml:21]: The latest Flutter stable version is `3.10.1` according to the [release doc](https://docs.flutter.dev/release/archive?tab=macos). Should we update this field?

**Review by manh-t — APPROVED** [review]
LGTM 🚀

File: .github/workflows/generate_sample.yml (hunk 7 above)
Comment (luongvo) [reply .github/workflows/generate_sample.yml:21]: @nmint8m It does not belong to this PR scope. We might consider updating in the following https://github.com/nimblehq/flutter-templates/pull/201 🙏

**Comment by luongvo** [comment]
@markgravity @nmint8m please have your final reviews on this PR soon 🤗

**Review by markgravity — APPROVED** [review]
✅

File: .github/workflows/generate_sample.yml [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+      - name: Remove the sample project
```
Comment (doannimble) [inline .github/workflows/generate_sample.yml:23]:
```suggestion
      - name: Remove the old sample project
```

**Review by doannimble — COMMENTED** [review]
rest lgtm 👍

**Review by nmint8m — APPROVED** [review]

File: .github/workflows/generate_sample.yml (hunk 8 above)
Comment (luongvo) [reply .github/workflows/generate_sample.yml]: Fixed in https://github.com/nimblehq/flutter-templates/pull/196/commits/d4f707779738c3a4bcab3c7f0b47f609d804b0ef

**Review by doannimble — APPROVED** [review]
Good to go now 🏃‍♂️

**Pull request merged**